include/zephyr/drivers/sensor.h

```diff
@@ -450,60 +411,120 @@ struct sensor_decoder_api {
 
 	/**
-	 * @brief Get the shift count of a particular channel (multiplier)
+	 * @brief Decode up to @p max_count samples from the buffer
 	 *
-	 * This value can be used by shifting the q31_t value resulting in the SI unit of the
-	 * reading. It is guaranteed that the shift for a channel will not change between frames.
+	 * Decode samples of channel @ref sensor_channel across multiple frames. If there exist
+	 * multiple instances of the same channel, @p channel_index is used to differentiate them.
+	 * As an example, assume a sensor provides 2 distance measurements:
 	 *
-	 * @param[in]  buffer The buffer provided on the @ref rtio context.
-	 * @param[in]  channel_type The c:enum:`sensor_channel` to query
-	 * @param[out] shift The bit shift of the channel for this data buffer.
-	 * @return 0 on success
-	 * @return -EINVAL if the @p channel_type doesn't exist in the buffer
-	 * @return <0 on error
-	 */
-	int (*get_shift)(const uint8_t *buffer, enum sensor_channel channel_type, int8_t *shift);
-
-	/**
-	 * @brief Decode up to N samples from the buffer
+	 * @code{.c}
+	 * // Decode the first channel instance of 'distance'
+	 * decoder->decode(buffer, SENSOR_CHAN_DISTANCE, 0, &fit, 5, out);
+	 * ...
 	 *
-	 * This function will never wrap frames. If 1 channel is available in the current frame and
-	 * @p max_count is 2, only 1 channel will be decoded and the frame iterator will be modified
-	 * so that the next call to decode will begin at the next frame.
+	 * // Decode the second channel instance of 'distance'
+	 * decoder->decode(buffer, SENSOR_CHAN_DISTANCE, 1, &fit, 5, out);
+	 * @endcode
 	 *
 	 * @param[in]     buffer The buffer provided on the @ref rtio context
+	 * @param[in]     channel The channel to decode
+	 * @param[in]     channel_idx The index of the channel
 	 * @param[in,out] fit The current frame iterator
-	 * @param[in,out] cit The current channel iterator
-	 * @param[out]    channels The channels that were decoded
-	 * @param[out]    values The scaled data that was decoded
 	 * @param[in]     max_count The maximum number of channels to decode.
-	 * @retval        > 0 The number of decoded values
-	 * @retval        0 Nothing else to decode on the @p buffer
-	 * @retval        < 0 Error
+	 * @param[out]    data_out The decoded data
+	 * @return 0 no more samples to decode
+	 * @return >0 the number of decoded frames
+	 * @return <0 on error
 	 */
-	int (*decode)(const uint8_t *buffer, sensor_frame_iterator_t *fit,
-		      sensor_channel_iterator_t *cit, enum sensor_channel *channels, q31_t *values,
-		      uint8_t max_count);
+	int (*decode)(const uint8_t *buffer, enum sensor_channel channel, size_t channel_idx,
+		      uint32_t *fit, uint16_t max_count, void *data_out);
 };
 
+/**
+ * @brief Used for iterating over the data frames via the sensor_decoder_api.
+ *
+ * Example usage:
+ *
+ * @code(.c)
+ *     struct sensor_decode_context ctx = SENSOR_DECODE_CONTEXT_INIT(
```

Comment:
looks great

Reply:
Sorry @teburd , I missed a few `s/sensor_frame_iterator_t/uint32_t/`. Should be fixed now.